### ISSUE TITLE
Add 'webglcontextchanged' event

### ISSUE THREAD

**Issue by grorg** [KhronosGroup/WebGL#2300]
When an implementation swaps GPUs the various limits and values exposed to the developer may change. They should be made aware of this, in case they've cached the values.

I suggest a new WebGLContextEvent with name 'webglcontextchanged'. It's basically indicating to the developer that something about the context is different. They don't have to do anything in response, but this gives them the opportunity.

If people agree I'll file a PR. My suggested wording would be along the lines of "fired when any detectable attribute for the WebGL context is changed by the user agent".

And maybe an example that shows max texture size or something changing.

**Comment by kenrussell** [comment]
This seems redundant. The webglcontextlost and webglcontextrestored events should already provide enough information to the developer that they need to re-cache any GPU related information. We could expand the documentation around these events, in particular webglcontextrestored, but I don't think we should add a new event type.

**Comment by grorg** [comment]
We originally considered just using contextrestored for this, sending such an event without a corresponding contextlostevent. But we were concerned that the majority of content in the wild was written to treat contextrestored as the time to reset everything. That's not what we want.

A GPU switch might happen quite often, and to a lot of content, no matter whether it requested low or high power. Forcing a real lost context at this time would be bad - you'd get churn on basically every running WebGL instance. 

Also, we don't think a GPU switch will cause many issues. Hopefully "it just works". What we're trying to provide is some way for content that really understands it is pushing the limits to just get a notification. It might be that it doesn't need to react to it. Most content won't.

**Comment by grorg** [comment]
During the WebGL conference call, @kenrussell asked for a list of situations where a GPU switch might not "work" (on macOS).

I'll try to gather more exact data, but at the moment we don't have anything official documenting it.

At the moment we think that WebGL 1 will never cause an error or crash that warrants a context lost. There might be cases where incorrect rendering is possible, such as if the point size was set to something bigger than GL_POINT_SIZE_MAX before the switch. Maybe that warrants a lost context, but it seems to be overkill - even if they restore the context they can't exceed that point size.

WebGL 2 might be more problematic. We haven't done a full review yet. Clearly there are a lot more limits to check, but in most cases the driver will just clamp to its max.

**Comment by kenrussell** [comment]
It's unappealing to rush in changes to this area of the spec. They'll stick around forever.

Point size is not a context state in OpenGL ES 2.0. Rather, the vertex shader writes the gl_PointSize variable if it wishes to change a particular point's size. Perhaps there are other arguments to be made for properties which change on the OpenGL context, but this isn't one of them.

The approach I'm considering in Chromium is to drop the hold on the discrete GPU if the tab is backgrounded, but to reactivate the discrete GPU if the tab's foregrounded again. In my thinking, this should prevent the application from needing to know about any GPU switches that occur -- regardless of what the underlying semantics really are -- in 99% of the cases, since they shouldn't be rendering when the tab is backgrounded anyway.

Apple's automatic graphics switching is remarkable. No other OS implements such seamless transitions. However, this also implies that this proposed webglcontextchanged event would only be useful on one OS and a small segment of machines, and the proper due diligence hasn't been done to understand whether it's really needed in the long run.

**Comment by grorg** [comment]
Your approach works for tabs that have WebGL content that requested high-power. You're guaranteeing they are always on the dGPU while in the foreground.

But it still may occur with WebGL content that requests low-power. It may be granted the dGPU because another tab has requested high-power. As that 2nd tab closes, the original tab moves back to the iGPU. Unless you mean that you're intending to keep all contexts on the dGPU if any one of them ever has been on it?

As for the general case, we don't think there will be much chance of a bad situation happening for WebGL 1. We're not sure about WebGL 2 yet.

Are you recommending that we lose/restore a context if we suspect there might be a problem? Since, unfortunately, we can't be sure yet, this would cause unnecessary churn. Our other alternative is just to let the content suffer without any way to detect. That's why we like the idea of a new event. It won't be used often, but it does give content a way to work around such issues.

**Comment by kenrussell** [comment]
Yes, rather than incorporating a spec change which may be regrettable in the long term, I'm suggesting that implementations that want to be conservative force a context loss and restoration if they think there will be problems with keeping contexts alive across GPU switches on macOS.

We really need to get clarity from Apple's OpenGL team on exactly what happens to the OpenGL context during GPU switches. Do limits change? Supported texture and renderbuffer formats? Supported extensions? If we had that clarity I would feel more comfortable specifying this. As it stands, this event would only be supported on one OS and developers would have no idea what to do in response to it.

**Comment by grorg** [comment]
> Do limits change? Supported texture and renderbuffer formats? Supported extensions?

Yes, they all may change.

**Comment by legion80** [comment]
@kenrussell, you state that an implementation that wants to be conservative could force context loss. I don't agree that not being conservative means not giving the web page any indication at all that something happened.

A WebGL site may want to proactively render a simpler scene if it knew that the mux occurred. One possible recommendation is that handlers of this new event re-query GPU limits, and make adjustments to their rendered scene.

**Comment by kenrussell** [comment]
@grorg @legion80 if anything "confusing" happens in response to the GPU switch -- for example, a previously-allocated texture's contents are lost because the texture format isn't supported on the other GPU -- then the only reasonable thing to do is lose the entire context. There's already an event for that and applications are already expected to respond to it.

Similarly, if the set of extensions changes in a way such that a previously-enabled WebGL extension can no longer be supported, the only reasonable thing to do is lose the context.

Both of these having been said, I agree that notifying the developer about the GPU change is reasonable, as long as the semantics of the new event are reasonable.

If you're willing to spec "webglcontextchanged" with semantics like:

"Notifies the application that the GPU used to render the WebGL scene has changed. Values such context-dependent limits may have changed and should be re-queried. All previously allocated WebGLObjects continue to be valid."

That would sound OK. Would you submit a pull request for discussion?

**Comment by grorg** [comment]
Thanks @kenrussell - I'll send a PR. 

And yes, if anything confusing happens we will lose the context. The OpenGL team says "things that can change are hardware features (specifically extensions), limits (like max point and/or line size, and also the list of supported compressed texture formats."

I don't think there are WebGL extensions that would be different between iGPU and dGPU on Apple hardware, other than compressed textures. It should be possible to remember if a compressed texture is being used and decide to lose the context, so I don't think that applies.

So, for WebGL 1, it seems like it is just the limits that would be troublesome. We could pick a more explicit name like "webglcontextlimitschanged". That way only content that knows it is going beyond the minimum requirements will have to listen. What do you think?

**Comment by legion80** [comment]
I still vote for `webglcontextchanged`. I'm concerned we would otherwise run into a similar situation with WebGL 2 and need to either rename this event, or add a new one.

**Comment by kenrussell** [comment]
"webglcontextchanged" sounds better to me too. As long as the event is non-destructive, it would be better to be able to expand its semantics in the future.

**Comment by RafaelCintron** [comment]
I agree with @kenrussell with regard to the non-destructive nature of the event as well as avoiding "confusion" with regard to unsupported texture formats and previously enabled extensions.

Even limits can be problematic.  If the previous max-texture-size limit was 4096 and the new one is 2048, WebGL contexts that contained textures that were larger than 2048 should also be come lost.

As part of #2283, we decided that we will ignore the powerPreference if no event listeners exist which would handle both "webglcontextlost" and "webglcontextrestored.  Are we going to adopt the same policy for this event?  In other words, if we evaluate outstanding resources/extensions and determine that we need to lose the context to avoid confusion, would be do it even if there were no "webglcontextlost" handlers present?  (I could go either way on this one)

**Comment by kenrussell** [comment]
I'm guessing that browsers will have to prioritize their own power management over the possibility of breaking applications. If there's an application out there which doesn't register a webglcontextlost handler, that probably can't be allowed to prevent the browser from making either an integrated -> discrete or discrete -> integrated GPU switch. Doing so would result in strange logic for the browser's "default" powerPreference -- whatever policy is chosen there -- and would probably prevent the browser from running WebGL content on the low-power GPU.

So, to answer your question above: yes, I think a context lost event should be dispatched in that scenario -- where the GPU switch is destructive, even if the application hasn't registered a webglcontextlost event handler.

**Comment by dmikis** [comment]
There're applications out there those in case `webglcontextlost` happens just abandon GL altogether and switch to some alternative way of presenting context to a user (and don't wait for `webglcontextrestore`). Should we specify that context became lost because of GPU switching in, let's say, `statusMessage` property of an event object? So such an application can decide what to do: wait for restoring its context or use fallbacks.

**Comment by kenrussell** [comment]
@dmikis fair point.

It would be better to define a boolean field on WebGLContextEvent like "forGPUSwitch" or "gpuSwitched" or "switchedGPU" rather than define the contents of the statusMessage.

**Comment by grorg** [comment]
@dmikis @kenrussell 

+1 to adding some fields to the WebGLContextEvent to provide a little more info (over status message).

**Comment by kenrussell** [comment]
@grorg do you want to propose spec text for these changes (context changed event, additional field in WebGLContextEvent)? Would you like me to?

Thinking about it more it would be better to add an enum defining the "reason" in WebGLContextEvent. (Not necessarily context lost, if WebGLContextEvent is also used for webglcontextchanged.)

**Comment by legion80** [comment]
> Thinking about it more it would be better to add an enum defining the "reason" in WebGLContextEvent.

Agreed. New reasons should extend the enum, not add more fields to the event.

**Comment by dmikis** [comment]
Agree with `reason` prop. For example, it may take `gpuswitched` (or `limitschanged`) value in case of GPU switching. Also may be handy to add `outofmemory` value to indicate lose of WebGL context in an event of exhausting memory resources.

**Comment by greggman** [comment]
I don't understand how this is going to work. So I have have a texture that's okay on high-per GPU but not ok on low-power GPU how does this event help?

Basically you're going to have to track if any feature is being used that won't work on the other GPU and if you switch then, if there are any resources that don't fit lose the context, if not change the context!??!?

Worse, since you the UA and switch at anytime then this can happen

     var maxSize = gl.getParameter(gl.MAX_TEXTURE_SIZE);
     gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, maxSize, 1, ...);
     for (x = 0; x < 1000000; ++x) { ... }

Because GL is asynchronous that call (which might be sitting in a command buffer because no gl.flush yet) could fail.

Basically a weglcontextchanged event is problematic. Anything since the time the event was queue to the time the event was delivered could fail and WebGL would no way to know that

To put it another way

1. webglcontextchanged gets queued
2. JS executes webglcontextchanged listener
3. JS asks for feature that doesn't exist on new GPU.

In that case you consider the app to have done something wrong. They should have noticed the webglcontextchanged event and not requested the non existent feature

vs

1. JS asks for feature that doesn't exist on new GPU.
2. That command as not executed yet
3. Browser switches GPUs
4. Command gets executed.

JavaScript had no chance to respond. Gets unexpected error.

I don't see how you can safely deliver a `webglcontextchanged` event.

This is one reason why `webglcontextrestored` *IS NOT ALLOWED HAPPEN* unless the app sets `event.preventDefault()` inside a `weblcontextlost` event. It's to let the app signal that it can actually handle it and to give it a chance to cancel all pending allocations, commands, BEFORE the the browser is allowed to make a new context.
